File: preinstall/sbr-config/internal/persistence/networkmanager.go
package persistence

import (
	"fmt"
	"log"
	"os"
	"path/filepath"
	"strings"

	"github.com/weka/tools/preinstall/sbr-config/internal/constants"
	"github.com/weka/tools/preinstall/sbr-config/internal/exec"
	"github.com/weka/tools/preinstall/sbr-config/internal/models"
)

// NetworkManagerBackend writes a NetworkManager dispatcher script for SBR
// persistence. Creates /etc/NetworkManager/dispatcher.d/50-sbr-config which
// is called by NetworkManager when interfaces come up or go down.
type NetworkManagerBackend struct{}

func (b *NetworkManagerBackend) WriteConfig(
	interfaces []models.InterfaceInfo,
	tables []models.RoutingTable,
) ([]string, error) {
	scriptPath := filepath.Join(constants.NMDispatcherDir, constants.NMDispatcherScript)

	if err := os.MkdirAll(constants.NMDispatcherDir, 0755); err != nil {
		return nil, fmt.Errorf("create NM dispatcher dir: %w", err)
	}

	script := b.generateScript(interfaces, tables)
	if err := exec.WriteFileAtomic(scriptPath, script, 0755); err != nil {
		return nil, fmt.Errorf("write NM dispatcher script: %w", err)
	}

	log.Printf("[INFO] Wrote NM dispatcher script: %s", scriptPath)
	return []string{scriptPath}, nil
}

func (b *NetworkManagerBackend) RemoveConfig() []string {
	scriptPath := filepath.Join(constants.NMDispatcherDir, constants.NMDispatcherScript)
	content, err := exec.ReadFile(scriptPath)
	if err != nil || content == "" {
		return nil
	}
	if strings.Contains(content, constants.ManagedComment) {
		os.Remove(scriptPath)
		log.Printf("[INFO] Removed NM dispatcher script: %s", scriptPath)
		return []string{scriptPath}
	}
	return nil
}

func (b *NetworkManagerBackend) Describe() string {
	scriptPath := filepath.Join(constants.NMDispatcherDir, constants.NMDispatcherScript)
	return fmt.Sprintf(
		"NetworkManager dispatcher script at %s\n"+
			"Called automatically when interfaces come up/down.", scriptPath)
}

func (b *NetworkManagerBackend) generateScript(
	interfaces []models.InterfaceInfo,
	tables []models.RoutingTable,
) string {
	// Build table name->number mapping.
	tableNum := make(map[string]int)
	for _, t := range tables {
		tableNum[t.Name] = t.Number
	}

	lines := []string{
		"#!/bin/bash",
		constants.ManagedComment,
		"# NetworkManager dispatcher script for source-based routing.",
		"# Called with $1=interface_name $2=action (up/down)",
		"",
		`IFACE="$1"`,
		`ACTION="$2"`,
		"",
		`case "$IFACE" in`,
	}

	for _, iface := range interfaces {
		tableName := constants.TableNamePrefix + iface.Name
		if _, ok := tableNum[tableName]; !ok {
			continue
		}

		// Build up commands from desired state (not delta).
		upCmds := []string{
			fmt.Sprintf("ip route replace %s dev %s src %s table %s",
				iface.Subnet, iface.Name, iface.IPAddress, tableName),
		}
		if iface.Gateway != nil {
			upCmds = append(upCmds, fmt.Sprintf(
				"ip route replace default via %s dev %s table %s",
				*iface.Gateway, iface.Name, tableName))
		}
		upCmds = append(upCmds, fmt.Sprintf(
			"ip rule add from %s table %s 2>/dev/null",
			iface.IPAddress, tableName))

		downCmds := []string{
			fmt.Sprintf("ip rule del from %s table %s", iface.IPAddress, tableName),
			fmt.Sprintf("ip route flush table %s", tableName),
		}

		lines = append(lines, fmt.Sprintf("    %s)", iface.Name))
		lines = append(lines, `        if [ "$ACTION" = "up" ]; then`)
		for _, cmd := range upCmds {
			lines = append(lines, fmt.Sprintf("            %s", cmd))
		}
		lines = append(lines, `        elif [ "$ACTION" = "down" ]; then`)
		for _, cmd := range downCmds {
			lines = append(lines, fmt.Sprintf("            %s 2>/dev/null", cmd))
		}
		lines = append(lines, "        fi")
		lines = append(lines, "        ;;")
	}

	lines = append(lines, "esac")
	lines = append(lines, "")

	return strings.Join(lines, "\n")
}

File: preinstall/sbr-config/sbr_config/persistence_backends/networkmanager.py
"""NetworkManager dispatcher script persistence backend."""

import logging
import os
from typing import List

from ..constants import (
    MANAGED_COMMENT,
    NM_DISPATCHER_DIR,
    NM_DISPATCHER_SCRIPT,
    TABLE_NAME_PREFIX,
)
from ..models import InterfaceInfo, PlannedChange, RoutingTable
from ..utils import command_exists, read_file, run_command, write_file_atomic
from .base import PersistenceBackend, group_by_name

logger = logging.getLogger(__name__)


class NetworkManagerBackend(PersistenceBackend):
    """Write a NetworkManager dispatcher script for SBR persistence.

    Creates /etc/NetworkManager/dispatcher.d/50-sbr-config which is
    called by NetworkManager when interfaces come up or go down.
    """

    def write_config(
        self,
        interfaces: List[InterfaceInfo],
        tables: List[RoutingTable],
        changes: List[PlannedChange],
    ) -> List[str]:
        script_path = os.path.join(NM_DISPATCHER_DIR, NM_DISPATCHER_SCRIPT)

        if not os.path.isdir(NM_DISPATCHER_DIR):
            os.makedirs(NM_DISPATCHER_DIR, exist_ok=True)

        # Build the dispatcher script
        script = self._generate_script(interfaces, tables, changes)
        write_file_atomic(script_path, script, mode=0o755)

        logger.info("Wrote NM dispatcher script: %s", script_path)

        # The dispatcher only fires for NM-managed devices. Hand-configured
        # storage NICs are often NM_CONTROLLED=no -- warn, or this
        # persistence silently does nothing for exactly those interfaces.
        self._warn_unmanaged(interfaces)

        return [script_path]

    def _warn_unmanaged(self, interfaces: List[InterfaceInfo]) -> None:
        if not command_exists("nmcli"):
            return
        for name, _ in group_by_name(interfaces):
            result = run_command(
                f"nmcli -t -f GENERAL.STATE device show {name}", check=False
            )
            if result.returncode == 0 and "unmanaged" in result.stdout:
                logger.warning(
                    "%s is unmanaged by NetworkManager; the dispatcher "
                    "script never fires for it, so SBR will not be restored "
                    "on link changes or reboot for this interface.",
                    name,
                )

    def remove_config(self) -> List[str]:
        script_path = os.path.join(NM_DISPATCHER_DIR, NM_DISPATCHER_SCRIPT)
        removed = []
        if os.path.exists(script_path):
            content = read_file(script_path)
            if content and MANAGED_COMMENT in content:
                os.unlink(script_path)
                removed.append(script_path)
                logger.info("Removed NM dispatcher script: %s", script_path)
        return removed

    def describe(self) -> str:
        script_path = os.path.join(NM_DISPATCHER_DIR, NM_DISPATCHER_SCRIPT)
        return (
            f"NetworkManager dispatcher script at {script_path}\n"
            f"Called automatically when interfaces come up/down."
        )

    def _generate_script(
        self,
        interfaces: List[InterfaceInfo],
        tables: List[RoutingTable],
        changes: List[PlannedChange],
    ) -> str:
        """Generate the bash dispatcher script content.

        Commands are derived from the desired interface state (not from
        the delta of this run) so the script always contains the COMPLETE
        set of commands needed after a reboot.
        """
        table_num = {t.name: t.number for t in tables}

        lines = [
            "#!/bin/bash",
            MANAGED_COMMENT,
            "# NetworkManager dispatcher script for source-based routing.",
            "# Called with $1=interface_name $2=action (up/down)",
            "",
            'IFACE="$1"',
            'ACTION="$2"',
            "",
            'case "$IFACE" in',
        ]

        for name, entries in group_by_name(interfaces):
            table_name = f"{TABLE_NAME_PREFIX}{name}"
            tnum = table_num.get(table_name)
            if tnum is None:
                continue

            # Derive the full command set from the desired state, covering
            # every address on the interface
            up_commands = []
            seen_subnets = set()
            for entry in entries:
                if entry.subnet in seen_subnets:
                    continue
                seen_subnets.add(entry.subnet)
                up_commands.append(
                    f"ip route replace {entry.subnet} dev {name} "
                    f"src {entry.ip_address} table {table_name}"
                )
            gateway = next((e.gateway for e in entries if e.gateway is not None), None)
            if gateway is not None:
                up_commands.append(
                    f"ip route replace default via {gateway} "
                    f"dev {name} table {table_name}"
                )
            for entry in entries:
                up_commands.append(
                    f"ip rule add from {entry.ip_address} table {table_name} 2>/dev/null"
                )

            down_commands = [
                f"ip rule del from {entry.ip_address} table {table_name}"
                for entry in entries
            ]
            down_commands.append(f"ip route flush table {table_name}")

            lines.append(f"    {name})")
            lines.append('        if [ "$ACTION" = "up" ]; then')
            for cmd in up_commands:
                lines.append(f"            {cmd}")
            lines.append('        elif [ "$ACTION" = "down" ]; then')
            for cmd in down_commands:
                lines.append(f"            {cmd} 2>/dev/null")
            lines.append("        fi")
            lines.append("        ;;")

        lines.append("esac")
        lines.append("")

        return "\n".join(lines)
